### ISSUE TITLE
Always exports the last trial rather than the best performing model

### ISSUE THREAD

**Issue by Unemyr** [keras-team/autokeras#998]
### Bug Description
API export_model() always seem to export the most recent trial model rather than the best performing model.

### Reproducing Steps
  * Step 1: Define a test run and determine a number X of max_trials (using trial and error) where the last run do not give you the best performing result 
  * Step 2: Export the model using export_model()
  * Step 3: Run a fit() of the exported model
  * Step 4: Observe how the model trained seem to have been exported from the last trial, rather than the best performing model.

### Expected Behavior
Should be exporting the best performing model based on the objective selected in fit().

### Setup Details
 - OS type and version: Ubuntu 19.10
 - Python: 3.7.5
 - autokeras: 1.02 (but I observe the same results in 1.0.0 and 1.0.1 as well)
 - scikit-learn: 0.22.1
 - numpy: 1.9.1
 - keras: 2.3.1
 - scipy: 0.14
 - tensorflow: 2.1.0
 - torch: 1.4.0

### Other Details
Refer to attached file for sample code snippet and log output observing the issue.

[autokeras_bug_report.txt](https://github.com/keras-team/autokeras/files/4239053/autokeras_bug_report.txt)

**Comment by haifeng-jin** [comment]
It looks like it is returning the last. Actually it is returning the best after fully training at the end of the search.

**Comment by Unemyr** [comment]
Hi Haifeng, just to clarify the X number of trials (20 in this case) does finish before doing the final fit. But it cannot be the case that the last trial that gives only NaN is the best performing model when the previous models (although not with great accuracy values) gives better results? Or am I somehow missing something, or using the APIs in the wrong way? Appreciate any suggestions, thank you!

**Comment by haifeng-jin** [comment]
I have never seen this before.
To my understanding, the issue is mainly related to export_model.

The behavior is highly likely that there is an exported preprocessing layer that is not adapted, which is a known bug.
However, I don't think we include any preprocessing layer in ImageRegressor.

Would you try best_model.summary() to see what is inside?
As long as it is an OK Keras model, it should have some loss value instead of NaN.

**Comment by Unemyr** [comment]
Hi Haifeng,
It looks like this (on the exact same code and repeatable training set as the original log file attached):

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_1 (InputLayer)            [(None, 18, 10, 1)]  0                                            
__________________________________________________________________________________________________
normalization (Normalization)   (None, 18, 10, 1)    3           input_1[0][0]                    
__________________________________________________________________________________________________
conv2d (Conv2D)                 (None, 9, 5, 128)    1280        normalization[0][0]              
__________________________________________________________________________________________________
separable_conv2d (SeparableConv (None, 9, 5, 128)    17664       conv2d[0][0]                     
__________________________________________________________________________________________________
separable_conv2d_1 (SeparableCo (None, 9, 5, 128)    17664       separable_conv2d[0][0]           
__________________________________________________________________________________________________
add (Add)                       (None, 9, 5, 128)    0           separable_conv2d_1[0][0]         
                                                                 conv2d[0][0]                     
__________________________________________________________________________________________________
separable_conv2d_2 (SeparableCo (None, 9, 5, 128)    17664       add[0][0]                        
__________________________________________________________________________________________________
separable_conv2d_3 (SeparableCo (None, 9, 5, 128)    17664       separable_conv2d_2[0][0]         
__________________________________________________________________________________________________
add_1 (Add)                     (None, 9, 5, 128)    0           separable_conv2d_3[0][0]         
                                                                 add[0][0]                        
__________________________________________________________________________________________________
separable_conv2d_4 (SeparableCo (None, 9, 5, 128)    17664       add_1[0][0]                      
__________________________________________________________________________________________________
separable_conv2d_5 (SeparableCo (None, 9, 5, 128)    17664       separable_conv2d_4[0][0]         
__________________________________________________________________________________________________
add_2 (Add)                     (None, 9, 5, 128)    0           separable_conv2d_5[0][0]         
                                                                 add_1[0][0]                      
__________________________________________________________________________________________________
separable_conv2d_6 (SeparableCo (None, 9, 5, 128)    17664       add_2[0][0]                      
__________________________________________________________________________________________________
separable_conv2d_7 (SeparableCo (None, 9, 5, 128)    17664       separable_conv2d_6[0][0]         
__________________________________________________________________________________________________
add_3 (Add)                     (None, 9, 5, 128)    0           separable_conv2d_7[0][0]         
                                                                 add_2[0][0]                      
__________________________________________________________________________________________________
separable_conv2d_8 (SeparableCo (None, 9, 5, 128)    17664       add_3[0][0]                      
__________________________________________________________________________________________________
separable_conv2d_9 (SeparableCo (None, 9, 5, 128)    17664       separable_conv2d_8[0][0]         
__________________________________________________________________________________________________
add_4 (Add)                     (None, 9, 5, 128)    0           separable_conv2d_9[0][0]         
                                                                 add_3[0][0]                      
__________________________________________________________________________________________________
separable_conv2d_10 (SeparableC (None, 9, 5, 256)    34176       add_4[0][0]                      
__________________________________________________________________________________________________
separable_conv2d_11 (SeparableC (None, 9, 5, 256)    68096       separable_conv2d_10[0][0]        
__________________________________________________________________________________________________
max_pooling2d (MaxPooling2D)    (None, 5, 3, 256)    0           separable_conv2d_11[0][0]        
__________________________________________________________________________________________________
conv2d_1 (Conv2D)               (None, 5, 3, 256)    33024       add_4[0][0]                      
__________________________________________________________________________________________________
add_5 (Add)                     (None, 5, 3, 256)    0           max_pooling2d[0][0]              
                                                                 conv2d_1[0][0]                   
__________________________________________________________________________________________________
flatten (Flatten)               (None, 3840)         0           add_5[0][0]                      
__________________________________________________________________________________________________
regression_head_1 (Dense)       (None, 1)            3841        flatten[0][0]                    
==================================================================================================
Total params: 317,060
Trainable params: 317,057
Non-trainable params: 3
__________________________________________________________________________________________________
Train on 3283 samples, validate on 1408 samples
Epoch 1/1000

  32/3283 [..............................] - ETA: 55s - loss: nan - mean_squared_error: nan

**Comment by haifeng-jin** [comment]
I think I know why.

It might because the normalization layer has no weights when exported.
Exported model's Normalization is load from disk but its weights are not saved to disk.
If the normalization is giving invalid output, the model loss will be nan.

You can just uninstall tensorflow and install the latest tf-nightly.
It should solve the problem.

**Comment by Unemyr** [comment]
Hi Haifeng,
I did try with tf-nightly and it does remove the occurence of NaN trials. Thanks for higlighting this - I guess all current development is following tf-nightly. I will go ahead and close this.

**Comment by LucaUrbinati44** [comment]
autokeras                 1.0.12
keras-tuner               1.0.2
tensorflow                2.4.1 

I have the same issue, that is export_model() returns the architecture of the best model with the weights of the last epoch instead of the best epoch.

However I think the problem is in keras-tuner because I found a [line](https://github.com/keras-team/keras-tuner/blob/86cff68ae4c7f35bb6557e75ec6a60e0766454d2/kerastuner/engine/tuner.py#L184) in the `save_model()` method of keras-team/keras-tuner/blob/master/kerastuner/engine/tuner.py that says:

`# TODO: save the top epoch checkpoints instead of last ones.`

I fixed it by saving the complete model architecture in `_checkpoint_model()` in keras-team/keras-tuner/blob/master/kerastuner/engine/tuner.py with this code:

```
def _checkpoint_model(self, model, trial_id, epoch):
        fname = self._get_checkpoint_fname(trial_id, epoch)
        # Save in TF format.
        model.save_weights(fname)
        path = self._get_checkpoint_dir(trial_id, epoch) # NEW LINE
        model.save(path+'/model') # NEW LINE
        return fname
```

Then in my main code I load the best saved model with `tensorflow.keras.models.load_model()`, after finding the best trial id and the best epoch.

I hope my experience can help someone else.
